sdk/go/common/workspace/plugins.go

```diff
@@ -1005,99 +1006,151 @@ type PluginSpec struct {
+	}
+	return spec, nil
+}
+
+func parsePluginSpecFromURL(
+	ctx context.Context, spec string, kind apitype.PluginKind,
+) (PluginSpec, parsePluginSpecInference, error) {
+	// Parse the version if available.  This can either be a simple semver version, or a git commit hash.
+	var version *semver.Version
+	var inference parsePluginSpecInference
+	spec, err := inference.parseVersion(spec, func(versionStr string) error {
+		v, err := semver.ParseTolerant(versionStr)
+		if err == nil {
 			version = &v
-		} else {
-			// Allow sha1 and sha256 hashes.
-			gitCommitRegex := regexp.MustCompile(`^[0-9a-fA-F]{4,64}$`)
-			if !gitCommitRegex.MatchString(versionStr) {
-				return PluginSpec{}, fmt.Errorf("VERSION must be valid semver or git commit hash: %s", versionStr)
-			}
+			return nil
+		}
+		if gitCommitRegex().MatchString(versionStr) {
 			version = &semver.Version{
 				// VersionStr cannot start with a 0, so we prefix it with an 'x' to avoid this.
 				Pre: []semver.PRVersion{{VersionStr: "x" + versionStr}},
 			}
+			return nil
 		}
+		return fmt.Errorf("VERSION must be valid semver or git commit hash: %s", versionStr)
+	})
+	if err != nil {
+		return PluginSpec{}, inference, err
 	}
 
+	parsedURL, err := url.Parse(spec)
+	if err != nil {
+		return PluginSpec{}, inference, fmt.Errorf("invalid URL: %w", err)
+	}
+	switch parsedURL.Scheme {
+	case "git", "https", "":
+		parsedURL.Scheme = "https"
+	default:
+		return PluginSpec{}, inference, errors.New(`unknown URL scheme: expected "git" or "https"`)
+	}
+
+	gitURL, _, err := gitutil.ParseGitRepoURL(parsedURL.String() + ".git")
+	if err != nil {
+		return PluginSpec{}, inference, err
+	}
+	pluginSpec := PluginSpec{
+		Name:    strings.ReplaceAll(strings.TrimPrefix(gitURL, "https://"), "/", "_"),
+		Kind:    kind,
+		Version: version,
+		// Prefix the url with `git://`, so we can later recognize this as a git URL.
+		PluginDownloadURL: func(url url.URL) string { url.Scheme = "git"; return url.String() }(*parsedURL),
+	}
+	inference.explicitPluginDownloadURL = true
+
+	// If there is no version specified, we version the plugin ourselves. This way the user gets
+	// a consistent experience once the plugin is installed, and won't have any problems when the repo
+	// is updated.  The version will then be added to the plugins SDK, and will be reused when the NewPluginSpec
+	// is used, so the user gets a consistent experience.
+	if pluginSpec.Version == nil {
+		var err error
+		pluginSpec.Version, err = gitutil.GetLatestTagOrHash(ctx, gitURL)
+		if err != nil {
+			return pluginSpec, inference, PluginVersionNotFoundError(err)
+		}
+	}
+
+	return pluginSpec, inference, nil
+}
+
+func parsePluginSpecFromName(
+	_ context.Context, spec string, kind apitype.PluginKind,
+) (PluginSpec, parsePluginSpecInference, error) {
+	var version *semver.Version
+	var inference parsePluginSpecInference
+	// Parse the version if available.  This can either be a simple semver version, or a git commit hash.
```

Comment:
Hrm for named plugins this shouldn't be a git commit hash I think, we only support semver versions for those?  I think that's what's happening below, but the comment is wrong.